File: engine/openscript/library/history.py
"""Immutable contributions with bounded append copying and cheap checkpoints."""

from dataclasses import dataclass
from typing import Optional, Sequence

from .values import Value

_WIDTH = 32


@dataclass(frozen=True, slots=True)
class _Chunk:
    values: tuple[Value, ...]
    prior: Optional["_Chunk"]


@dataclass(frozen=True, slots=True)
class ContributionHistory:
    """A version shares its sealed prefix with previous versions and forks."""

    head: Optional[_Chunk] = None
    tail: tuple[Value, ...] = ()
    count: int = 0
    maximum: int = 0
    seed_count: int = 0

    def append(self, value: Value, length: Optional[int]) -> "ContributionHistory":
        full = len(self.tail) == _WIDTH
        head = _Chunk(self.tail, self.head) if full else self.head
        tail = (value,) if full else self.tail + (value,)
        maximum = max(self.maximum, length or 0)
        # Recursive seeds retain their previous bounded-buffer availability.
        seed_count = min(self.seed_count + 1, max(maximum, 1))
        return ContributionHistory(head, tail, self.count + 1, maximum, seed_count)

    def __deepcopy__(self, memo):
        """All reachable state is immutable, so checkpoint copying shares it."""
        return self

    def view(self, length: int) -> "ContributionView":
        size = min(length, self.count)
        chunks = []
        remaining = size - len(self.tail)
        head = self.head
        while remaining > 0 and head is not None:
            chunks.append(head.values)
            head = head.prior
            remaining -= _WIDTH
        return ContributionView(self, tuple(chunks), size)


@dataclass(frozen=True, slots=True)
class ContributionView(Sequence[Value]):
    """A transient oldest-first sequence indexing only the requested chunks."""

    history: ContributionHistory
    chunks: tuple[tuple[Value, ...], ...]
    size: int

    def __len__(self):
        return self.size

    def __getitem__(self, index):
        if isinstance(index, slice):
            return [self[at] for at in range(*index.indices(self.size))]
        if index < 0:
            index += self.size
        if index < 0 or index >= self.size:
            raise IndexError(index)
        back = self.size - 1 - index
        tail = self.history.tail
        if back < len(tail):
            return tail[len(tail) - 1 - back]
        before = back - len(tail)
        return self.chunks[before // _WIDTH][_WIDTH - 1 - before % _WIDTH]

File: src/core/stdlib/values/history.ts
/** Immutable contribution chunks, shared by state checkpoints and independent forks. */
import type { Value } from './value.js';

const WIDTH = 32;

interface Chunk {
  readonly values: readonly Value[];
  readonly prior: Chunk | null;
}

/** Appending copies at most one short tail, never the contributed prefix. */
export class ContributionHistory {
  readonly head: Chunk | null;
  readonly tail: readonly Value[];
  readonly count: number;

  constructor(
    head: Chunk | null = null,
    tail: readonly Value[] = Object.freeze([]),
    count = 0,
  ) {
    this.head = head;
    this.tail = tail;
    this.count = count;
    Object.freeze(this);
  }

  append(value: Value): ContributionHistory {
    const full = this.tail.length === WIDTH;
    const head = full ? Object.freeze({ values: this.tail, prior: this.head }) : this.head;
    const tail = Object.freeze(full ? [value] : [...this.tail, value]);
    return new ContributionHistory(head, tail, this.count + 1);
  }

  /** Index chunk references once; each subsequent newest-first read is constant work. */
  view(length: number): ContributionView {
    const size = Math.min(length, this.count);
    const chunks: (readonly Value[])[] = [];
    let remaining = size - this.tail.length;
    let head = this.head;
    while (remaining > 0 && head !== null) {
      chunks.push(head.values);
      head = head.prior;
      remaining -= WIDTH;
    }
    return new ContributionView(this.tail, chunks, size);
  }
}

/** A transient index into the current window. It is not retained in state. */
export class ContributionView {
  private readonly tail: readonly Value[];
  private readonly chunks: readonly (readonly Value[])[];
  readonly size: number;

  constructor(
    tail: readonly Value[], chunks: readonly (readonly Value[])[], size: number,
  ) {
    this.tail = tail;
    this.chunks = chunks;
    this.size = size;
  }

  at(back: number): Value {
    if (back < 0 || back >= this.size || !Number.isInteger(back)) return null;
    if (back < this.tail.length) return this.tail[this.tail.length - 1 - back]!;
    const before = back - this.tail.length;
    return this.chunks[Math.floor(before / WIDTH)]![WIDTH - 1 - before % WIDTH]!;
  }
}
